Fixes #2687 - Leases file is no longer being parsed on every request

dmitri-d opened the pull request:
This is work in progress and currently only isc dhcpd running on linux is supported (support for BSD is coming). Please ignore the failing tests, these are due to changes in internal API.

dmitri-d committed:
Fixes #2687: Leases file is no longer parsed on every request

  Also: refs 12392, #12359, #12425, #1090

lzap commented:
I see some huge refactoring in this PR, is this expected?

dmitri-d commented:
This is based on work in https://github.com/theforeman/smart-proxy/pull/390 PR. There's also a commit for https://github.com/theforeman/smart-proxy/pull/406 PR, I might rebase this PR to exclude it. You can ignore first two commits if you are only interested in dhcp changes.

timogoebel commented:
@witlessbird : I'd be happy to test this. What changes/commits do I need to apply agains 1.11.1?

dmitri-d commented:
@timogoebel: thanks a bunch! The easiest would be to use this branch -- all commits were made against develop (not 1.11), and the changes in the last commit depend on the changes in the first.

timogoebel commented:
@witlessbird : I tested this, but somehow it does not work 100%.

With your patch:

```
D, [2016-04-29T16:56:33.773647 #482] DEBUG -- : Loading subnets for 127.0.0.1
D, [2016-04-29T16:56:33.773888 #482] DEBUG -- : Loading subnet data for 172.23.9.0/255.255.255.0
E, [2016-04-29T16:56:33.774181 #482] ERROR -- : Record 172.23.9.0/172.23.9.49 not found
D, [2016-04-29T16:56:33.774222 #482] DEBUG -- : Record 172.23.9.0/172.23.9.49 not found
```

Without your patch, it finds the entry:

```
D, [2016-04-29T16:53:42.366182 #32624] DEBUG -- : Added a reservation: 172.23.9.49:00:50:XX:XX:XX:XX:XXXXX
```

In addition foreman_dhcp_brower Plugin shows the subnet being empty with your patch (it's not without).
How can I help with debugging this?

dmitri-d commented:
@timogoebel: thanks for testing this! 

To confirm: you are using isc dhcp (these changes are not relevant for other providers. Other providers can be broken in this branch atm). If so, you should see a list of lease and host records that were processed on startup, does it include the ones you search for?

dmitri-d commented:
@timogoebel: also, check there are no errors on startup related to dhcp...

timogoebel commented:
@witlessbird 
Yes, it initializes everything correctly and finds the subnet on startup (and I do use isc-dhcpd)

```
I, [2016-05-02T08:33:50.924345 #7394]  INFO -- : Successfully initialized 'tftp'
D, [2016-05-02T08:33:50.925335 #7394] DEBUG -- : Added a subnet: 172.23.9.0
[...]
I, [2016-05-02T08:33:50.985827 #7394]  INFO -- : Successfully initialized 'dhcp_isc'
I, [2016-05-02T08:33:50.986187 #7394]  INFO -- : Successfully initialized 'dhcp'
```

dmitri-d commented:
@timogoebel: you should see something similar to:

```
D, [2016-05-03T06:08:13.500534 #21612] DEBUG -- : Added a subnet: 192.168.42.0
D, [2016-05-03T06:08:13.501780 #21612] DEBUG -- : Added a reservation: 192.168.42.193:52:54:00:8e:5f:08:foreman-test
D, [2016-05-03T06:08:13.501837 #21612] DEBUG -- : Added a reservation: 192.168.42.173:e8:2a:ea:14:88:ef:lucid-nonsense
D, [2016-05-03T06:08:13.501881 #21612] DEBUG -- : Added a reservation: 192.168.42.191:52:54:00:03:00:d8:mco-agent
D, [2016-05-03T06:08:13.501925 #21612] DEBUG -- : Added a reservation: 192.168.42.190:52:54:00:e8:2c:78:mco-master
D, [2016-05-03T06:08:13.501968 #21612] DEBUG -- : Added a reservation: 192.168.42.5:60:a4:4c:d3:b3:e0:space-monster
D, [2016-05-03T06:08:13.502013 #21612] DEBUG -- : Added a reservation: 192.168.42.2:00:24:36:9a:21:b2:stranger-here-myself
D, [2016-05-03T06:08:13.502056 #21612] DEBUG -- : Added a reservation: 192.168.42.171:c8:bc:c8:9d:d0:c5:unwitting-accomplice-lan
D, [2016-05-03T06:08:13.502097 #21612] DEBUG -- : Added a reservation: 192.168.42.170:c8:bc:c8:e3:8a:e6:unwitting-accomplice
D, [2016-05-03T06:08:13.517242 #21612] DEBUG -- : Added a lease record: 192.168.42.240:7c:d1:c3:f2:02:a3
D, [2016-05-03T06:08:13.517353 #21612] DEBUG -- : Added a lease record: 192.168.42.245:58:55:ca:d6:28:2f
D, [2016-05-03T06:08:13.517402 #21612] DEBUG -- : Added a lease record: 192.168.42.227:6c:40:08:e8:18:2b
D, [2016-05-03T06:08:13.517444 #21612] DEBUG -- : Added a lease record: 192.168.42.246:ac:bc:32:ae:86:37
D, [2016-05-03T06:08:13.517485 #21612] DEBUG -- : Added a lease record: 192.168.42.250:d0:03:4b:82:02:f6
D, [2016-05-03T06:08:13.517525 #21612] DEBUG -- : Added a lease record: 192.168.42.223:b8:8d:12:57:f0:13
D, [2016-05-03T06:08:13.517573 #21612] DEBUG -- : Added a lease record: 192.168.42.247:64:bc:0c:48:9d:64
D, [2016-05-03T06:08:13.517614 #21612] DEBUG -- : Added a lease record: 192.168.42.213:a4:77:33:a9:05:d6
D, [2016-05-03T06:08:13.517654 #21612] DEBUG -- : Added a lease record: 192.168.42.242:00:1b:67:07:95:1c
D, [2016-05-03T06:08:13.517701 #21612] DEBUG -- : Deleted a lease record: 192.168.42.247:64:bc:0c:48:9d:64
D, [2016-05-03T06:08:13.517757 #21612] DEBUG -- : Added a lease record: 192.168.42.247:64:bc:0c:48:9d:64
D, [2016-05-03T06:08:13.517797 #21612] DEBUG -- : Deleted a lease record: 192.168.42.213:a4:77:33:a9:05:d6
D, [2016-05-03T06:08:13.517834 #21612] DEBUG -- : Added a lease record: 192.168.42.213:a4:77:33:a9:05:d6
D, [2016-05-03T06:08:13.517871 #21612] DEBUG -- : Deleted a lease record: 192.168.42.242:00:1b:67:07:95:1c
D, [2016-05-03T06:08:13.517907 #21612] DEBUG -- : Added a lease record: 192.168.42.242:00:1b:67:07:95:1c
I, [2016-05-03T06:08:13.517989 #21612]  INFO -- : Successfully initialized 'dhcp_isc'

```

Note a bunch of host- and lease-records being added (and deleted). Do you see anything like this? If not, can you check that dhcp_isc configuration is pointing to the correct (as in actually used by isc dhcpd) leases file?

dmitri-d commented:
@timogoebel: additionally, you should periodically (depends on how busy your dhcpd is) see something similar to:

```
D, [2016-05-03T06:13:02.137977 #21612] DEBUG -- : caught :modify event on /var/lib/dhcpd/dhcpd.leases.
D, [2016-05-03T06:13:02.138993 #21612] DEBUG -- : Deleted a lease record: 192.168.42.242:00:1b:67:07:95:1c
D, [2016-05-03T06:13:02.139123 #21612] DEBUG -- : Added a lease record: 192.168.42.242:00:1b:67:07:95:1c
"dhcpd.leases:access"
```

dmitri-d commented:
Fixed a bunch (not all) of tests. @timogoebel: I may have fixed the issue you were seeing. Out of curiosity -- are you running smart-proxy under ruby 1.8.7?

timogoebel commented:
@witlessbird : Thanks. If I find the time today, I'll retest the latest version and let you know how it works. 

I'm using default ruby on rhel 7.2.

``` bash
$ ruby --version
ruby 2.0.0p598 (2014-11-13) [x86_64-linux]
```

dmitri-d commented:
Fixed a bunch more tests; expanded test coverage of dhcp/dhcp_api.rb (api controller)

dmitri-d commented:
All tests should be passing now. Added di wirings tests.

timogoebel commented:
@witlessbird : I retested your latest commit. This looks very good. Free IP suggesting and dhcp browser work fine. Smart Proxy catches modify events on the lease file. Everything feels faster. 👍 

I also tried to provision a VM but did not succeed.
Foreman shows:

```
Create DHCP Settings for tst-timo-11.lxsbx.ka.de.dm-drogeriemarkt.com task failed with the following error: ERF12-6899 [ProxyAPI::ProxyException]: Unab
le to set DHCP entry ([RestClient::BadRequest]: 400 Bad Request) for proxy https://foremansp.ka.de.dm-drogeriemarkt.com:8443/dhcp
 | ProxyAPI::ProxyException: ERF12-6899 [ProxyAPI::ProxyException]: Unable to set DHCP entry ([RestClient::BadRequest]: 400 Bad Request) for proxy https://foremansp.ka.de.dm-drogeri
emarkt.com:8443/dhcp
 | /usr/share/foreman/lib/proxy_api/dhcp.rb:66:in `rescue in set'
 | /usr/share/foreman/lib/proxy_api/dhcp.rb:62:in `set'
 | /usr/share/foreman/lib/net/dhcp/record.rb:29:in `create'
 | /usr/share/foreman/app/models/concerns/orchestration/dhcp.rb:51:in `set_dhcp'
```

Now to the interesting part: Smart Proxy says:

```
D, [2016-05-06T08:49:23.332744 #22444] DEBUG -- : Loading subnets for 127.0.0.1
D, [2016-05-06T08:49:23.332898 #22444] DEBUG -- : Loading subnet data for 172.23.9.0/255.255.255.0
D, [2016-05-06T08:49:23.336708 #22444] DEBUG -- : omshell: executed - set name = "tst-timo-11.lxsbx.ka.de.dm-drogeriemarkt.com"
D, [2016-05-06T08:49:23.336786 #22444] DEBUG -- : nil
D, [2016-05-06T08:49:23.336838 #22444] DEBUG -- : omshell: executed - set ip-address = 172.23.9.186
D, [2016-05-06T08:49:23.336870 #22444] DEBUG -- : nil
D, [2016-05-06T08:49:23.336918 #22444] DEBUG -- : omshell: executed - set hardware-address = 00:50:56:84:ac:1a
D, [2016-05-06T08:49:23.336949 #22444] DEBUG -- : nil
D, [2016-05-06T08:49:23.336983 #22444] DEBUG -- : omshell: executed - set hardware-type = 1
D, [2016-05-06T08:49:23.337011 #22444] DEBUG -- : nil
E, [2016-05-06T08:49:23.337187 #22444] ERROR -- : undefined method `bootServer' for 127.0.0.1:Proxy::DHCP::ISC::Provider
D, [2016-05-06T08:49:23.337225 #22444] DEBUG -- : undefined method `bootServer' for 127.0.0.1:Proxy::DHCP::ISC::Provider (NoMethodError)
/var/tmp/dhcp-parse/modules/dhcp_isc/dhcp_isc_main.rb:43:in `add_record'
/var/tmp/dhcp-parse/modules/dhcp/dhcp_api.rb:88:in `block in <class:DhcpApi>'
/usr/local/share/gems/gems/sinatra-1.4.6/lib/sinatra/base.rb:1610:in `call'
/usr/local/share/gems/gems/sinatra-1.4.6/lib/sinatra/base.rb:1610:in `block in compile!'
```

dmitri-d commented:
rebased, fixed the issue discovered by @timogoebel, added tests.

dmitri-d commented:
rebased.

dmitri-d commented:
@mmoll: now with support for bsd (kqueue-based lease file monitoring). Atm I don't have a bsd-based machine to test this with, could you give it a try?

dmitri-d commented:
rebased.

dmitri-d commented:
rebased.

dmitri-d commented:
rebased.

dmitri-d commented:
[test]

mmoll commented:
I finally tested this with 1.12.0 + this patch on FreeBSD:

```
D, [2016-07-12T21:08:25.329541 #98927] DEBUG -- : 'dynflow' settings: 'console_auth': true (default), 'database': /var/lib/foreman-proxy/dynflow/dynflow.sqlite (default), 'enabled': true
D, [2016-07-12T21:08:25.332789 #98927] DEBUG -- : 'ssh' settings: 'enabled': true, 'local_working_dir': /var/tmp (default), 'remote_working_dir': /var/tmp (default), 'ssh_identity_key_file': /usr/local/share/foreman-proxy/.ssh/id_rsa_foreman_proxy, 'ssh_user': root (default)
D, [2016-07-12T21:08:25.336051 #98927] DEBUG -- : 'salt' settings: 'api_auth': pam, 'api_password': password, 'api_url': https://localhost:8080, 'api_username': saltuser, 'autosign_file': /etc/salt/autosign.conf (default), 'enabled': true, 'salt_command_user': root (default), 'use_api': false (default)
D, [2016-07-12T21:08:25.342053 #98927] DEBUG -- : 'tftp' settings: 'enabled': true, 'tftp_servername': 192.168.200.11, 'tftproot': /netboot
D, [2016-07-12T21:08:25.347194 #98927] DEBUG -- : 'dhcp' settings: 'enabled': true, 'server': 127.0.0.1 (default), 'subnets': [] (default), 'use_provider': dhcp_isc (default)
D, [2016-07-12T21:08:25.348306 #98927] DEBUG -- : Providers ['dhcp_isc'] are going to be configured for 'dhcp'
D, [2016-07-12T21:08:25.362379 #98927] DEBUG -- : 'dhcp_isc' settings: 'config': /usr/local/etc/dhcpd.conf, 'key_name': omapi_key, 'key_secret': Zf+vpmcbi92dXuD6tVbNLdNA+Yajc9fyRPwGLTgquRGefnn7NwznKAO0x4nWSSq4aRc6jRFWx4MQThDwPe4Sqg==, 'leases': /var/db/dhcpd/dhcpd.leases, 'leases_file_observer': kqueue_leases_file_observer, 'omapi_port': 7911 (default), 'server': 127.0.0.1, 'subnets': [], 'use_provider': dhcp_isc
I, [2016-07-12T21:08:25.362810 #98927]  INFO -- : Successfully initialized 'dynflow'
I, [2016-07-12T21:08:25.439083 #98927]  INFO -- : Successfully initialized 'ssh'
I, [2016-07-12T21:08:25.439230 #98927]  INFO -- : Successfully initialized 'salt'
I, [2016-07-12T21:08:25.439298 #98927]  INFO -- : Successfully initialized 'foreman_proxy'
I, [2016-07-12T21:08:25.439359 #98927]  INFO -- : Successfully initialized 'tftp'
D, [2016-07-12T21:08:25.479545 #98927] DEBUG -- : trying to find an ip address, we got {:from=>"192.168.200.50", :to=>"192.168.200.59"}
D, [2016-07-12T21:08:25.479959 #98927] DEBUG -- : Added a subnet: 192.168.200.0
E, [2016-07-12T21:08:25.480914 #98927] ERROR -- : Couldn't enable 'dhcp_isc': Invalid MAC 
D, [2016-07-12T21:08:25.480961 #98927] DEBUG -- : ["/usr/local/share/foreman-proxy/lib/proxy/validations.rb:23:in `validate_mac'", "/usr/local/share/foreman-proxy/modules/dhcp_common/record.rb:16:in `initialize'", "/usr/local/share/foreman-proxy/modules/dhcp_common/record/reservation.rb:10:in `initialize'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_file_parser.rb:40:in `new'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_file_parser.rb:40:in `block in parse_config_and_leases_for_records'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_file_parser.rb:24:in `scan'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_file_parser.rb:24:in `parse_config_and_leases_for_records'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:76:in `update_subnet_data'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:26:in `block in monitor_started'", "/usr/local/share/foreman-proxy/modules/dhcp_common/subnet_service.rb:140:in `block in group_changes'", "/usr/local/lib/ruby/2.2/monitor.rb:211:in `mon_synchronize'", "/usr/local/share/foreman-proxy/modules/dhcp_common/subnet_service.rb:140:in `group_changes'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:20:in `monitor_started'", "/usr/local/share/foreman-proxy/modules/dhcp_isc/kqueue_leases_file_observer.rb:34:in `start'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:228:in `block in start_services'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:226:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:226:in `start_services'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:211:in `configure_plugin'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:89:in `block in configure'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:89:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:89:in `configure'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `block in initialize_plugins'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `initialize_plugins'", "/usr/local/share/foreman-proxy/lib/launcher.rb:122:in `launch'", "/usr/local/sbin/smart-proxy:7:in `<main>'"]
E, [2016-07-12T21:08:25.481146 #98927] ERROR -- : Error during startup, terminating. undefined method `stop' for nil:NilClass
D, [2016-07-12T21:08:25.481184 #98927] DEBUG -- : ["/usr/local/share/foreman-proxy/modules/dhcp_isc/kqueue_leases_file_observer.rb:39:in `stop'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:110:in `block (2 levels) in stop_services'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:110:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:110:in `block in stop_services'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:109:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:109:in `stop_services'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:92:in `rescue in configure'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:88:in `configure'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `block in initialize_plugins'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `each'", "/usr/local/share/foreman-proxy/lib/proxy/plugin_initializer.rb:160:in `initialize_plugins'", "/usr/local/share/foreman-proxy/lib/launcher.rb:122:in `launch'", "/usr/local/sbin/smart-proxy:7:in `<main>'"]
```

dmitri-d commented:
@mmoll: fixed the issue you discovered.

dmitri-d commented:
@mmoll: any word?

mmoll commented:
Yes, it dosen't work. It seems that things get written, but when being read, the old state is getting into the proxy, I didn't dig further, because of ENOTIME. I think the leases file is not getting watched correctly, so the proxy doesn't get that it did change.

dmitri-d commented:
@mmoll: could you give this another try? I changed how I handle file deletes, which should do the trick.

mmoll commented:
@witlessbird sems to work fine for me :)

mmoll commented:
@witlessbird this needs a rebase now :)

dmitri-d commented:
rebased.

timogoebel commented:
Let me quickly sum up the status of this. I tested this on EL7, @mmoll tested this on FreeBSD. I think we just need to find somebody who knows the smart proxy code well can review this PR. Any volunteers?

domcleal commented:
I plan to review it soon if it hasn't been done by then, not forgotten about it.

dmitri-d commented:
[test]

dmitri-d commented:
test failure has been fixed in https://github.com/theforeman/smart-proxy/pull/442.

dmitri-d commented:
rebased, should fix test failures.

domcleal commented on modules/dhcp/dhcp_api.rb:
```diff
@@ -102,15 +101,10 @@ def load_subnet_data
+      content_type :json
       record = server.find_record(@subnet.network, params[:record])
       log_halt 404, "Record #{params[:network]}/#{params[:record]} not found" unless record
       server.del_record @subnet, record
```
I'm unsure if this is correct with the native_ms provider - it looks like `del_record` will return the string response from the `execute` call to delete the record. Both the libvirt and ISC providers look like they're going to return `nil`, but the execute response won't be valid JSON if that's what's returned.

dmitri-d replied:
Agreed, responses to DELETE shouldn't have body anyway.

domcleal commented on lib/proxy/memory_store.rb:
```diff
@@ -21,6 +21,10 @@ def values(*path)
       path.empty? ? get_all_values(@root) : get_all_values(get(@root, path))
     end
 
+    def clear
```
Also add a test to test/memory_store_test.rb

domcleal commented:
Thanks, it generally looks very good. My only main concern is that the new Proxy::DHCP::ISC::IscStateChangesObserver class doesn't appear to have any test coverage, but it could do with some.

I think we can live without coverage of the platform-specific observers, unless you'd like to mock those too (which is probably quite straight forward), but this contains a mix of logic for parsing includes out of files, plus adding, removing and updating records in the subnet service which is important to have coverage of.

dmitri-d commented:
Responded to feedback:
- added  test fo verify MemoryStore#clear method
- refactored IscStateChangesObserver to make it more testable and added tests (also see https://github.com/theforeman/smart-proxy/pull/409/files#diff-4599cc4165d8bc18a6d28dbb13e797feR1 and https://github.com/theforeman/smart-proxy/pull/409/files#diff-5d61320075bd97a660de5dfaad8d909fR1 and their tests)

> I think we can live without coverage of the platform-specific observers, unless you'd like to mock those too 

I promise I'll think about how to best test those -- it's a bit tricky as they essentially create an event loop. Can we continue without these tests for the time being please?

domcleal commented:
> > I think we can live without coverage of the platform-specific observers, unless you'd like to mock those too
> 
> I promise I'll think about how to best test those -- it's a bit tricky as they essentially create an event loop. Can we continue without these tests for the time being please?

Indeed, that's fine.

Thanks for the update, the testing is much improved. Subject to #444 fixing tests first, this is reviewed.

domcleal commented:
[test] with json_pure fixed.

domcleal commented on modules/dhcp_isc/leases_file.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+
+      def close
+        fd.close unless fd.nil?
```
I think `@fd` also has to be set to nil here, otherwise the old fd is used when `hosts_and_leases` is called again from `leases_recreated`:

<pre>
E, [2016-07-29T10:53:27.693723 #5691] ERROR -- : Error occured when monitoring /tmp/dhcp.leases
D, [2016-07-29T10:53:27.693750 #5691] DEBUG -- : closed stream (IOError)
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/leases_file.rb:13:in `read'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/leases_file.rb:13:in `hosts_and_leases'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:35:in `block in leases_recreated'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_common/subnet_service.rb:140:in `block in group_changes'
/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_common/subnet_service.rb:140:in `group_changes'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/isc_state_changes_observer.rb:27:in `leases_recreated'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:25:in `block (2 levels) in monitor_leases'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:17:in `each'
/home/dcleal/code/foreman/smart-proxy/modules/dhcp_isc/inotify_leases_file_observer.rb:17:in `block in monitor_leases'
</pre>

dmitri-d replied:
Duh!

dmitri-d commented:
- Proxy::DHCP::ISC::LeasesFile#close resets file descriptor now
- Added a test to verify the behavior above
- rebased

domcleal commented:
Merged as 7bd71b5, thanks @witlessbird.
